File: desktop/src/shared/protocol.ts
/**
 * CibilBazaar Dialer — shared protocol types (see /shared/protocol.md for spec).
 * Mirrors android/app/src/main/java/.../protocol/Protocol.kt exactly.
 */

export type MessageType =
  | "HELLO"
  | "HELLO_ACK"
  | "DIAL_REQUEST"
  | "DIAL_ACK"
  | "CALL_RESULT"
  | "CALL_RESULT_ACK"
  | "SMS_REQUEST"
  | "SMS_ACK"
  | "PING"
  | "PONG"
  | "ERROR";

export interface Envelope<T = unknown> {
  v: 1;
  type: MessageType;
  id: string;
  ts: number;
  payload: T;
}

export interface HelloPayload {
  deviceName: string;
  role: "ANDROID" | "DESKTOP";
  appVersion: string;
  pairingCode?: string;
}

export interface HelloAckPayload {
  accepted: boolean;
  deviceName: string;
  reason?: string;
}

export interface DialRequestPayload {
  rowId: string;
  mobile: string;
  name: string;
}

export interface DialAckPayload {
  rowId: string;
  opened: boolean;
}

export type CallStatus =
  | "CONNECTED"
  | "NO_ANSWER"
  | "BUSY"
  | "FAILED"
  | "REJECTED";

export interface CallResultPayload {
  rowId: string;
  mobile: string;
  durationSeconds: number;
  status: CallStatus;
  startedAtEpochMs: number;
  endedAtEpochMs: number;
}

export interface CallResultAckPayload {
  rowId: string;
  saved: boolean;
}

export interface ErrorPayload {
  code: string;
  message: string;
}

export interface SmsRequestPayload {
  rowId: string;
  mobile: string;
  message?: string; // optional pre-filled body (e.g. a follow-up template)
}

export interface SmsAckPayload {
  rowId: string;
  opened: boolean;
}

export const PROTOCOL_VERSION = 1 as const;
export const BLUETOOTH_SPP_UUID = "94f39d29-7d6d-437d-973b-fba39e49d4ee";
export const WIFI_TCP_PORT = 47521;
export const WIFI_DISCOVERY_UDP_PORT = 47522;
export const HEARTBEAT_INTERVAL_MS = 5000;
export const HEARTBEAT_TIMEOUT_MS = 15000;
export const RECONNECT_BACKOFF_MS = [1000, 2000, 4000, 8000, 16000, 30000];

export function makeEnvelope<T>(type: MessageType, payload: T): Envelope<T> {
  return {
    v: PROTOCOL_VERSION,
    type,
    id: cryptoRandomId(),
    ts: Date.now(),
    payload,
  };
}

export function cryptoRandomId(): string {
  // RFC4122-ish v4 without pulling in a dependency
  return "xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx".replace(/[xy]/g, (c) => {
    const r = (Math.random() * 16) | 0;
    const v = c === "x" ? r : (r & 0x3) | 0x8;
    return v.toString(16);
  });
}

export function encodeLine(envelope: Envelope<unknown>): string {
  return JSON.stringify(envelope) + "\n";
}

/**
 * Incrementally feeds raw bytes/strings from a stream-based transport
 * (Bluetooth RFCOMM / USB serial) and yields complete parsed envelopes as
 * they arrive, since those transports have no built-in message framing.
 */
export class LineFrameDecoder {
  private buffer = "";

  push(chunk: string): Envelope<any>[] {
    this.buffer += chunk;
    const out: Envelope<any>[] = [];
    let idx: number;
    while ((idx = this.buffer.indexOf("\n")) >= 0) {
      const line = this.buffer.slice(0, idx).trim();
      this.buffer = this.buffer.slice(idx + 1);
      if (!line) continue;
      try {
        out.push(JSON.parse(line));
      } catch {
        // malformed line — drop silently, keeps stream alive
      }
    }
    return out;
  }
}

/**
 * Common interface every transport (WiFi/Bluetooth/USB) implements so the
 * rest of the app (callEngine, UI) never needs to know which one is active.
 */
export interface Transport {
  readonly name: "WIFI" | "BLUETOOTH" | "USB";
  isConnected(): boolean;
  connect(): Promise<void>;
  disconnect(): Promise<void>;
  send(envelope: Envelope<unknown>): Promise<void>;
  onMessage(handler: (envelope: Envelope<any>) => void): void;
  onStatusChange(handler: (connected: boolean) => void): void;
}

File: android/app/src/main/java/com/cibilbazaar/dialer/protocol/Protocol.kt
package com.cibilbazaar.dialer.protocol

import org.json.JSONObject
import java.util.UUID

/**
 * CibilBazaar Dialer — wire protocol (see /shared/protocol.md).
 * Mirrors desktop/src/shared/protocol.ts exactly. Manual JSON parsing (no
 * reflection library) keeps this dependency-free and fast on-device.
 */

const val PROTOCOL_VERSION = 1
const val BLUETOOTH_SPP_UUID = "94f39d29-7d6d-437d-973b-fba39e49d4ee"
const val WIFI_TCP_PORT = 47521
const val WIFI_DISCOVERY_UDP_PORT = 47522
const val HEARTBEAT_INTERVAL_MS = 5000L
const val HEARTBEAT_TIMEOUT_MS = 15000L
val RECONNECT_BACKOFF_MS = longArrayOf(1000, 2000, 4000, 8000, 16000, 30000)

enum class MessageType {
    HELLO, HELLO_ACK, DIAL_REQUEST, DIAL_ACK, CALL_RESULT, CALL_RESULT_ACK, SMS_REQUEST, SMS_ACK, PING, PONG, ERROR
}

enum class CallStatus { CONNECTED, NO_ANSWER, BUSY, FAILED, REJECTED }

data class Envelope(
    val v: Int,
    val type: MessageType,
    val id: String,
    val ts: Long,
    val payload: JSONObject
) {
    fun toJson(): JSONObject = JSONObject()
        .put("v", v)
        .put("type", type.name)
        .put("id", id)
        .put("ts", ts)
        .put("payload", payload)

    fun encodeLine(): String = toJson().toString() + "\n"

    companion object {
        fun make(type: MessageType, payload: JSONObject): Envelope =
            Envelope(PROTOCOL_VERSION, type, UUID.randomUUID().toString(), System.currentTimeMillis(), payload)

        fun parse(line: String): Envelope? {
            return try {
                val obj = JSONObject(line)
                Envelope(
                    v = obj.optInt("v", 1),
                    type = MessageType.valueOf(obj.getString("type")),
                    id = obj.getString("id"),
                    ts = obj.optLong("ts", System.currentTimeMillis()),
                    payload = obj.optJSONObject("payload") ?: JSONObject()
                )
            } catch (e: Exception) {
                null // malformed line — drop silently, keep stream alive
            }
        }
    }
}

data class DialRequestPayload(val rowId: String, val mobile: String, val name: String) {
    companion object {
        fun fromJson(json: JSONObject) = DialRequestPayload(
            rowId = json.getString("rowId"),
            mobile = json.getString("mobile"),
            name = json.optString("name", "")
        )
    }
}

data class CallResultPayload(
    val rowId: String,
    val mobile: String,
    val durationSeconds: Int,
    val status: CallStatus,
    val startedAtEpochMs: Long,
    val endedAtEpochMs: Long
) {
    fun toJson(): JSONObject = JSONObject()
        .put("rowId", rowId)
        .put("mobile", mobile)
        .put("durationSeconds", durationSeconds)
        .put("status", status.name)
        .put("startedAtEpochMs", startedAtEpochMs)
        .put("endedAtEpochMs", endedAtEpochMs)
}

fun helloPayload(deviceName: String, pairingCode: String? = null): JSONObject {
    val obj = JSONObject()
        .put("deviceName", deviceName)
        .put("role", "ANDROID")
        .put("appVersion", "1.0.0")
    if (pairingCode != null) obj.put("pairingCode", pairingCode)
    return obj
}

fun dialAckPayload(rowId: String, opened: Boolean): JSONObject =
    JSONObject().put("rowId", rowId).put("opened", opened)

data class SmsRequestPayload(val rowId: String, val mobile: String, val message: String?) {
    companion object {
        fun fromJson(json: JSONObject) = SmsRequestPayload(
            rowId = json.getString("rowId"),
            mobile = json.getString("mobile"),
            message = if (json.has("message") && !json.isNull("message")) json.optString("message") else null
        )
    }
}

fun smsAckPayload(rowId: String, opened: Boolean): JSONObject =
    JSONObject().put("rowId", rowId).put("opened", opened)

/**
 * Incrementally feeds raw text chunks from a stream-based transport
 * (Bluetooth RFCOMM / USB serial) and yields complete parsed envelopes —
 * mirrors LineFrameDecoder in desktop/src/shared/protocol.ts.
 */
class LineFrameDecoder {
    private val buffer = StringBuilder()

    fun push(chunk: String): List<Envelope> {
        buffer.append(chunk)
        val out = mutableListOf<Envelope>()
        var idx: Int
        while (true) {
            idx = buffer.indexOf("\n")
            if (idx < 0) break
            val line = buffer.substring(0, idx).trim()
            buffer.delete(0, idx + 1)
            if (line.isEmpty()) continue
            Envelope.parse(line)?.let { out.add(it) }
        }
        return out
    }
}
